endpoints/openrtb2/auction.go

```diff
@@ -56,24 +62,56 @@ func (deps *endpointDeps) Auction(w http.ResponseWriter, r *http.Request, _ http
 	w.Write(responseBytes)
 }
 
-// parseRequest turns the HTTP request into an OpenRTB request.
+// parseRequest turns the HTTP request into an OpenRTB request. This is guaranteed to return:
 //
-// This will return an error if the request couldn't be parsed, or if the request isn't valid according
-// to the OpenRTB 2.5 spec.
+//   - A context which times out appropriately, given the request.
+//   - A cancellation function which should be called if the auction finishes early.
 //
-// It will also return errors for some of the "strong recommendations" in the spec, as long as
-// the same request can be sent in a better way which agrees with the recommendations.
-func (deps *endpointDeps) parseRequest(httpRequest *http.Request) (*openrtb.BidRequest, error) {
-	var ortbRequest openrtb.BidRequest
-	if err := json.NewDecoder(httpRequest.Body).Decode(&ortbRequest); err != nil {
-		return nil, err
+// If the errors list is empty, then the returned request will be valid according to the OpenRTB 2.5 spec.
+// In case of "strong recommendations" in the spec, it tends to be restrictive. If a better workaround is
+// possible, it will return errors with messages that suggest improvements.
+//
+// If the errors list has at least one element, then no guarantees are made about the returned request.
+func (deps *endpointDeps) parseRequest(httpRequest *http.Request) (req *openrtb.BidRequest, ctx context.Context, cancel func(), errs []error) {
```

Comment:
["The Context should be the _first parameter_, typically named ctx"](https://golang.org/pkg/context/)